File: lachancuoicgoi_flutter/lib/analysis/analysis_result.dart
import '../core/risk_level.dart';
import 'analysis_level.dart';

class KeywordMatch {
  const KeywordMatch({
    required this.keyword,
    required this.level,
    required this.category,
    this.startIndex = -1,
    this.endIndex = -1,
    this.isFuzzy = false,
  });

  final String keyword;
  final RiskLevel level;
  final String category;
  final int startIndex;
  final int endIndex;
  final bool isFuzzy;

  KeywordMatch copyWith({
    String? keyword,
    RiskLevel? level,
    String? category,
    int? startIndex,
    int? endIndex,
    bool? isFuzzy,
  }) {
    return KeywordMatch(
      keyword: keyword ?? this.keyword,
      level: level ?? this.level,
      category: category ?? this.category,
      startIndex: startIndex ?? this.startIndex,
      endIndex: endIndex ?? this.endIndex,
      isFuzzy: isFuzzy ?? this.isFuzzy,
    );
  }

  Map<String, Object?> toJson() {
    return {
      'keyword': keyword,
      'level': level.storageName,
      'category': category,
      'startIndex': startIndex,
      'endIndex': endIndex,
      'isFuzzy': isFuzzy,
    };
  }

  factory KeywordMatch.fromJson(Map<String, Object?> json) {
    return KeywordMatch(
      keyword: json['keyword'] as String? ?? '',
      level: RiskLevel.fromString(json['level'] as String?),
      category: json['category'] as String? ?? 'Unknown',
      startIndex: (json['startIndex'] as num?)?.toInt() ?? -1,
      endIndex: (json['endIndex'] as num?)?.toInt() ?? -1,
      isFuzzy: json['isFuzzy'] as bool? ?? false,
    );
  }

  @override
  bool operator ==(Object other) {
    return other is KeywordMatch &&
        other.keyword == keyword &&
        other.level == level &&
        other.category == category &&
        other.startIndex == startIndex &&
        other.endIndex == endIndex &&
        other.isFuzzy == isFuzzy;
  }

  @override
  int get hashCode => Object.hash(
        keyword,
        level,
        category,
        startIndex,
        endIndex,
        isFuzzy,
      );
}

class AnalysisResult {
  const AnalysisResult({
    required this.overallRiskLevel,
    required this.matches,
    this.reason,
    this.analysisLevel = AnalysisLevel.l1,
    this.alertEnabled = false,
    this.confidence = -1,
    this.modelName,
    this.isError = false,
  });

  final RiskLevel overallRiskLevel;
  final List<KeywordMatch> matches;
  final String? reason;
  final AnalysisLevel analysisLevel;
  final bool alertEnabled;
  final double confidence;
  final String? modelName;
  final bool isError;

  AnalysisResult copyWith({
    RiskLevel? overallRiskLevel,
    List<KeywordMatch>? matches,
    String? reason,
    AnalysisLevel? analysisLevel,
    bool? alertEnabled,
    double? confidence,
    String? modelName,
    bool? isError,
  }) {
    return AnalysisResult(
      overallRiskLevel: overallRiskLevel ?? this.overallRiskLevel,
      matches: matches ?? this.matches,
      reason: reason ?? this.reason,
      analysisLevel: analysisLevel ?? this.analysisLevel,
      alertEnabled: alertEnabled ?? this.alertEnabled,
      confidence: confidence ?? this.confidence,
      modelName: modelName ?? this.modelName,
      isError: isError ?? this.isError,
    );
  }

  Map<String, Object?> toJson() {
    return {
      'overallRiskLevel': overallRiskLevel.storageName,
      'matches': matches.map((match) => match.toJson()).toList(),
      'reason': reason,
      'analysisLevel': analysisLevel.id,
      'alertEnabled': alertEnabled,
      'confidence': confidence,
      'modelName': modelName,
      'isError': isError,
    };
  }

  factory AnalysisResult.fromJson(Map<String, Object?> json) {
    final rawMatches = json['matches'];
    return AnalysisResult(
      overallRiskLevel:
          RiskLevel.fromString(json['overallRiskLevel'] as String?),
      matches: rawMatches is List
          ? rawMatches
              .whereType<Map>()
              .map((match) =>
                  KeywordMatch.fromJson(match.cast<String, Object?>()))
              .toList()
          : const [],
      reason: json['reason'] as String?,
      analysisLevel: AnalysisLevel.fromId(json['analysisLevel'] as String?),
      alertEnabled: json['alertEnabled'] as bool? ?? false,
      confidence: (json['confidence'] as num?)?.toDouble() ?? -1,
      modelName: json['modelName'] as String?,
      isError: json['isError'] as bool? ?? false,
    );
  }
}

File: lachancuocgoi/app/src/main/java/com/example/lachancuocgoi/Analysis/AnalysisResult.kt
package com.example.lachancuocgoi.Analysis

import com.example.lachancuocgoi.RiskLevel

/**
 * Đại diện cho kết quả của một phiên phân tích văn bản.
 *
 * @param overallRiskLevel Mức độ rủi ro tổng thể được tính toán.
 * @param matches Danh sách các từ khóa đáng ngờ được tìm thấy.
 * @param reason Lý do cho kết quả phân tích (thường được sử dụng với phân tích AI).
 * @param analysisLevel Cấp độ phân tích tạo ra kết quả này ("L1", "L2", hoặc "L3").
 */
data class AnalysisResult(
    val overallRiskLevel: RiskLevel,
    val matches: List<KeywordMatch>,
    val reason: String? = null,
    val analysisLevel: AnalysisLevel = AnalysisLevel.L1,
    val alertEnabled: Boolean = true,  // NEW: Điều khiển alert (default true for L1 compatibility)
    val isError: Boolean = false,      // MỚI: Cờ xác định có lỗi mạng/API để chuyển chế độ sang L2
    val confidence: Float = -1f,       // -1 = chưa tính, 0.0-1.0 = độ tin cậy của engine
    val modelName: String? = null      // NEW: Lưu tên cụ thể của model AI đã dùng (nếu có)
)

/**
 * Đại diện cho một từ khóa đáng ngờ được tìm thấy trong văn bản.
 *
 * @param keyword Từ khóa gốc (chưa qua chuẩn hóa).
 * @param level Mức độ rủi ro của từ khóa.
 * @param category Danh mục của từ khóa (ví dụ: LỪA ĐẢO, BẠO LỰC).
 */
data class KeywordMatch(
    val keyword: String,
    val level: RiskLevel,
    val category: String,
    val startIndex: Int = -1, // Vị trí bắt đầu của từ khóa trong danh sách tokens
    val endIndex: Int = -1,   // Vị trí kết thúc của từ khóa
    val isFuzzy: Boolean = false // True nếu match qua fuzzy matching (Levenshtein)
)
